### PR TITLE
PRC-56 : Enter dob during create contact

### PR DESCRIPTION
**Pull request by james-baker-moj**
* Convert creating a contact to a journey and collapse some of the routes
 * Add page to enter DOB
 * Add support for session handling in tests
 * Add support for checking error messages when not linked to a field

### CODE REVIEW

File: server/routes/contacts/create/createContactMiddleware.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import { Request, RequestHandler } from 'express'
+
+const ensureInCreateContactJourney = (): RequestHandler => {
+  return async (req: Request<{ journeyId: string }, unknown, unknown>, res, next) => {
```
Comment (james-baker-moj) [inline server/routes/contacts/create/createContactMiddleware.ts:4]: I'm planning to add some constraints on the number of journeys as per CSIP but will do that in a separate PR.
Also need to think about the journey entry points, they probably need to be encoded in the URL somewhere to allow us to restart properly (or perhaps we just error?)

File: server/routes/contacts/create/enter-dob/createContactEnterDobSchemas.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+const MONTH_REQUIRED_MESSAGE = 'Enter month'
+const YEAR_REQUIRED_MESSAGE = 'Enter year'
+
+export const createContactEnterDobSchema = () => async () => {
+  return createSchema({
+    isDobKnown: z.string({ message: SELECT_IS_DOB_KNOWN_MESSAGE }).transform(value => value === 'true'),
+    day: z.coerce.number({ message: DAY_TYPE_MESSAGE }).optional(),
+    month: z.coerce.number({ message: MONTH_TYPE_MESSAGE }).optional(),
+    year: z.coerce.number({ message: YEAR_TYPE_MESSAGE }).optional(),
+    dob: z.date().optional(),
```
Comment (james-baker-moj) [inline server/routes/contacts/create/enter-dob/createContactEnterDobSchemas.ts:18]: Not actually optional but couldn't find a better way of doing this with zod. It's populated via superRefine and transform functions but if you don't populate it here then it's not present in the generated type.

File: server/views/pages/contacts/create/enterDob.njk [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+                </div>
+            </form>
+
+        </div>
+    </div>
+
+{% endblock %}
+{% block pageScripts %}
+    <script src="/js/card.js"></script>
+{% endblock %}
```
Comment (timharrison-moj) [inline server/views/pages/contacts/create/enterDob.njk:111]: Does it need the card.js on this page? Didn't see a card on it?

Comment (james-baker-moj) [reply server/views/pages/contacts/create/enterDob.njk:111]: nope, will tidy up and imports. good spot

File: server/@types/express/index.d.ts [hunk 4]
```diff
@@ -3,13 +3,16 @@ import { fieldErrors } from '../../middleware/validationMiddleware'
...
+
   interface SessionData {
     returnTo: string
     nowInMinutes: number
     prisonId: string
     prisonName: string
     search: string
     journey: object
+    createContactJourneys: Record<string, CreateContactJourney>
   }
```
Comment (timharrison-moj) [inline server/@types/express/index.d.ts:16]: Does it need both `journey` and the list of journey types in `createContactJourneys`? Or does the latter replace the former? Might be my vague understanding of this..

File: server/routes/contacts/create/createContactMiddleware.ts (hunk 1 above)
Comment (timharrison-moj) [reply server/routes/contacts/create/createContactMiddleware.ts:4]: If they're inside one journey, but paste in a URL, or type one, then yes I think we'd need to recognise this and ensure they can't jump into a later part of the same journey or a different journey altogether.

File: server/routes/contacts/create/createContactRoutes.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+import SuccessController from './success/createContactSuccessController'
+import logPageViewMiddleware from '../../../middleware/logPageViewMiddleware'
+import EnterNameController from './enter-name/createContactEnterNameController'
+import { validate } from '../../../middleware/validationMiddleware'
+import { createContactEnterNameSchemaFactory } from './enter-name/createContactEnterNameSchemas'
+import CreateContactEnterDobController from './enter-dob/createContactEnterDobController'
+import { createContactEnterDobSchema } from './enter-dob/createContactEnterDobSchemas'
+import StartCreateContactJourneyController from './start/startCreateContactJourneyController'
+import ensureInCreateContactJourney from './createContactMiddleware'
+
```
Comment (timharrison-moj) [inline server/routes/contacts/create/createContactRoutes.ts:12]: I like this - all of the routes together for creating a contact. This is more like BVLS method, so I find this more familiar and easier to follow.

File: server/routes/contacts/create/enter-dob/createContactEnterDobSchemas.ts (hunk 2 above)
Comment (timharrison-moj) [reply server/routes/contacts/create/enter-dob/createContactEnterDobSchemas.ts:18]: I'll need to stare at this for a while .. or read the zod docs..

**Review by timharrison-moj — APPROVED** [review]
LGTM. Couple of questions only, but a great start at journey setup.

File: server/@types/express/index.d.ts (hunk 4 above)
Comment (james-baker-moj) [reply server/@types/express/index.d.ts:16]: No it could be cleaned up, I left it there as wasn't sure who created it and what it was for.
